Library/Homebrew/utils/formatter.rb

```diff
@@ -0,0 +1,47 @@
+require "utils/tty"
+
+module Formatter
+  module_function
+
+  def arrow(string, color: nil)
+    if color
+      "#{Tty.send(color)}==>#{Tty.reset} #{string}"
+    else
+      "==> #{string}"
+    end
+  end
+
+  def headline(string, color: nil)
+    arrow("#{Tty.bold}#{string}#{Tty.reset}", color: color)
+  end
+
+  def success(string)
+    colorize_text_before_colon(string, :green)
+  end
+
+  def warning(string)
+    colorize_text_before_colon(string, :yellow)
+  end
+
+  def error(string)
+    colorize_text_before_colon(string, :red)
+  end
+
+  def url(string)
+    descenders_no_underline("#{Tty.blue.underline}#{string}#{Tty.reset}")
+  end
+
+  def colorize_text_before_colon(string, color)
+    label, message = string.scan(/(.*?:)?(.*)/m)[0]
+
+    if label.nil?
+      "#{Tty.send(color)}#{message}#{Tty.reset}"
+    else
+      "#{Tty.send(color).underline}#{descenders_no_underline(label)}#{Tty.reset}#{message}"
+    end
+  end
+
+  def descenders_no_underline(string)
+    string.gsub(/([-+:jgqpy])/, "#{Tty.no_underline}\\1#{Tty.underline}")
```

Comment:
Why is this excluding `-`, `+`, and `:` that are clearly characters that stay above the baseline? Think about the `:` in `Formatter.url("http://example.com")`; not underlining this colon looks really weird.

Aside from that, I'm really against this kind of micro-optimization, not to mention that it only looks decent for letters where the descender spans the width of the entire character, e.g. `g` in most fonts, but looks like something is missing for letters like `p`. The problem with the underline is not that it touches descenders but the underline itself that is undesirable from a typographical point of view. Furthermore, this manual treatment will defeat proper treatment by the font and/or the font rendering of my terminal (assuming it did have some clever handling of underlines like a word processor often has).

Finally, check out e.g. the `Security` link on the bottom of every GitHub page. If you hover over it, the whole text is underlined (including the `y`). Like everyone else, we shouldn't bother special-casing this.

Practical suggestions:

- Underline all characters of a URL unconditionally.
- Underline all characters of an error/warning label unconditionally or simply avoid underlining if this feels like it looks too ugly (and yes, underlines are always ugly).